packages/iov-ethereum/src/ethereumconnection.ts

```diff
@@ -138,6 +155,130 @@ export class EthereumConnection implements AtomicSwapConnection {
+    const senderBegin = swapIdEnd;
+    const senderEnd = senderBegin + 32;
+    const recipientBegin = senderEnd;
+    const recipientEnd = recipientBegin + 32;
+    const hashBegin = recipientEnd;
+    const hashEnd = hashBegin + 32;
+    const amountBegin = hashEnd;
+    const amountEnd = amountBegin + 32;
+    const timeoutBegin = amountEnd;
+    const timeoutEnd = timeoutBegin + 32;
+    const erc20ContractAddressBegin = timeoutEnd;
+    const erc20ContractAddressEnd = erc20ContractAddressBegin + 32;
+
+    const erc20Token =
+      prefix === SwapIdPrefix.Ether
+        ? null
+        : [...erc20Tokens.values()].find(
+            token =>
+              token.contractAddress.toLowerCase() ===
+              Abi.decodeAddress(
+                bytes.slice(erc20ContractAddressBegin, erc20ContractAddressEnd),
+              ).toLowerCase(),
+          );
+
+    if (prefix === SwapIdPrefix.Erc20 && !erc20Token) {
+      // Found swap for a token we’re not tracking
+      return null;
+    }
+
+    const fractionalDigits = erc20Token ? erc20Token.decimals : constants.primaryTokenFractionalDigits;
+    const tokenTicker = erc20Token ? (erc20Token.symbol as TokenTicker) : constants.primaryTokenTicker;
+    const amount = {
+      quantity: new BN(bytes.slice(amountBegin, amountEnd)).toString(),
+      fractionalDigits: fractionalDigits,
+      tokenTicker: tokenTicker,
+    };
+
+    return {
+      kind: SwapProcessState.Open,
+      data: {
+        id: {
+          prefix: prefix,
+          data: bytes.slice(swapIdBegin, swapIdEnd) as SwapIdBytes,
+        },
+        sender: toChecksummedAddress(Abi.decodeAddress(bytes.slice(senderBegin, senderEnd))),
+        recipient: toChecksummedAddress(Abi.decodeAddress(bytes.slice(recipientBegin, recipientEnd))),
+        hash: bytes.slice(hashBegin, hashEnd) as Hash,
+        amounts: [amount],
+        timeout: {
+          height: new BN(bytes.slice(timeoutBegin, timeoutEnd)).toNumber(),
+        },
+      },
+    };
+  }
+
+  private static parseClaimedEventBytes(
+    bytes: Uint8Array,
+  ): {
+    readonly kind: SwapProcessState.Claimed;
+    readonly swapIdBytes: SwapIdBytes;
+    readonly preimage: Preimage;
+  } {
+    const swapIdBegin = 0;
+    const swapIdEnd = swapIdBegin + 32;
+    const preimageBegin = swapIdEnd;
+    const preimageEnd = preimageBegin + 32;
+
+    return {
+      kind: SwapProcessState.Claimed,
+      swapIdBytes: bytes.slice(swapIdBegin, swapIdEnd) as SwapIdBytes,
+      preimage: bytes.slice(preimageBegin, preimageEnd) as Preimage,
+    };
+  }
+
+  private static parseAbortedEventBytes(
+    bytes: Uint8Array,
+  ): {
+    readonly kind: SwapProcessState.Aborted;
+    readonly swapIdBytes: SwapIdBytes;
+  } {
+    const swapIdBegin = 0;
+    const swapIdEnd = swapIdBegin + 32;
+
+    return {
+      kind: SwapProcessState.Aborted,
+      swapIdBytes: bytes.slice(swapIdBegin, swapIdEnd) as SwapIdBytes,
+    };
+  }
+
+  private static updateSwapInList(
```

Comment:
It looks like this does the same as BCP's AtomicSwapMerger. Could you have a look at that?

Reply:
So the difference is that `AtomicSwapMerger` is merging full transactions into a list of swaps, whereas what's happening here is that events from the smart contract are being merged into a list of swaps. Claimed/Aborted events have very minimal information compared to transactions.

Reply:
Perfect, thanks for clarifying